utility/src/main/java/org/oppia/util/parser/CustomTagHandler.kt

```diff
@@ -0,0 +1,158 @@
+  private val indent = context.resources.getDimensionPixelSize(R.dimen.bullet_leading_margin)
+  private val listItemIndent = indent * 2
+  private val bulletSpan = BulletSpanWithRadius(context, indent)
+
+  private val listParents: Stack<ListItemTag> = Stack<ListItemTag>()
+
+  override fun handleTag(
+    opening: Boolean,
+    tag: String,
+    output: Editable,
+    xmlReader: XMLReader?
+  ) {
+    when (tag) {
+      UL_TAG -> {
+        if (opening) {
+          listParents.push(UnorderedListTag())
+        } else {
+          listParents.pop()
+        }
+      }
+      OL_TAG -> {
+        if (opening) {
+          listParents.push(OrderedListTag())
+        } else {
+          listParents.pop()
+        }
+      }
+      LI_TAG -> {
+        try {
+          if (opening) {
+            listParents.peek().openItem(output)
+          } else {
+            listParents.peek().closeItem(output, listParents.size)
+          }
+        } catch (e: EmptyStackException) {
+          throw EmptyStackException()
+        }
+      }
+      else -> {
+        // Found an unsupported tag $tag
+      }
+    }
+  }
+
+  /** Abstract super class for [UnorderedListTag] and [OrderedListTag]. */
+  private abstract class ListItemTag {
+    /**
+     * Opens a new list item.
+     *
+     * @param text
+     */
+    open fun openItem(text: Editable) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
+        text.append("\n")
+      }
+      val len = text.length
+      text.setSpan(this, len, len, Spanned.SPAN_MARK_MARK)
+    }
+
+    /**
+     * Closes a list item.
+     *
+     * @param text
+     * @param indentation
+     */
+    fun closeItem(text: Editable, indentation: Int) {
+      if (text.length > 0 && text[text.length - 1] != '\n') {
+        text.append("\n")
+      }
+      val replaces = getReplaces(text, indentation)
+      val len = text.length
+      val listTag = getLast(text)
+      val where = text.getSpanStart(listTag)
+      text.removeSpan(listTag)
+      if (where != len) {
+        for (replace in replaces) {
+          text.setSpan(replace, where, len, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+        }
+      }
+    }
+
+    protected abstract fun getReplaces(text: Editable?, indentation: Int): Array<Any>
+    /** Note: This knows that the last returned object from getSpans() will be the most recently added. */
+    private fun getLast(text: Spanned): ListItemTag? {
+      val listTags = text.getSpans(0, text.length, ListItemTag::class.java)
+      return if (listTags.size == 0) {
+        null
+      } else listTags[listTags.size - 1]
+    }
+  }
+
+  /** Class representing the unordered list (`<ul>`) HTML tag. */
+  private inner class UnorderedListTag : ListItemTag() {
+    override fun getReplaces(
+      text: Editable?,
+      indentation: Int
+    ): Array<Any> {
+      var bulletMargin: Int = indent
+      if (indentation > 1) {
+        bulletMargin = indent - bulletSpan.getLeadingMargin(true)
```

Comment:
Could we instead create a new bullet span here, compute its leading margin as needed, and then return it as part of the array below? We're creating a span anyway, and this would avoid needing the class ``bulletSpan`` variable.